File: verso-api/src/models/Duel.model.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IDuel extends Document {
  _id: mongoose.Types.ObjectId;
  theme: string;
  challengerId: mongoose.Types.ObjectId;
  challengeeId: mongoose.Types.ObjectId;
  challengerPoemId: mongoose.Types.ObjectId;
  challengeePoemId: mongoose.Types.ObjectId | null;
  status: 'pending' | 'active' | 'completed' | 'declined';
  votesForChallenger: number;
  votesForChallengee: number;
  voterIds: mongoose.Types.ObjectId[];
  winnerId: mongoose.Types.ObjectId | null;
  endsAt: Date;
  createdAt: Date;
}

const DuelSchema = new Schema<IDuel>(
  {
    theme: { type: String, required: true, trim: true, maxlength: 100 },
    challengerId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    challengeeId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    challengerPoemId: { type: Schema.Types.ObjectId, ref: 'Poem', required: true },
    challengeePoemId: { type: Schema.Types.ObjectId, ref: 'Poem', default: null },
    status: { type: String, enum: ['pending', 'active', 'completed', 'declined'], default: 'pending', index: true },
    votesForChallenger: { type: Number, default: 0 },
    votesForChallengee: { type: Number, default: 0 },
    voterIds: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    winnerId: { type: Schema.Types.ObjectId, ref: 'User', default: null },
    endsAt: { type: Date, index: true },
  },
  { timestamps: true }
);

DuelSchema.index({ status: 1, endsAt: 1 });
DuelSchema.index({ challengerId: 1, createdAt: -1 });
DuelSchema.index({ challengeeId: 1, createdAt: -1 });

export const Duel = mongoose.model<IDuel>('Duel', DuelSchema);

File: verso/lib/shared/models/duel_model.dart
/// Duel model matching backend schema
class DuelModel {
  final String id;
  final String theme;
  final String challengerId;
  final String challengeeId;
  final String? challengerPoemId;
  final String? challengeePoemId;
  final String status;
  final int votesForChallenger;
  final int votesForChallengee;
  final int challengerPercent;
  final int challengeePercent;
  final int totalVotes;
  final List<String> voterIds;
  final String? winnerId;
  final DateTime? endsAt;
  final DateTime createdAt;

  const DuelModel({
    required this.id,
    required this.theme,
    required this.challengerId,
    required this.challengeeId,
    this.challengerPoemId,
    this.challengeePoemId,
    required this.status,
    required this.votesForChallenger,
    required this.votesForChallengee,
    required this.challengerPercent,
    required this.challengeePercent,
    required this.totalVotes,
    required this.voterIds,
    this.winnerId,
    this.endsAt,
    required this.createdAt,
  });

  factory DuelModel.fromJson(Map<String, dynamic> json) {
    return DuelModel(
      id: json['id'] as String,
      theme: json['theme'] as String,
      challengerId: json['challengerId'] as String,
      challengeeId: json['challengeeId'] as String,
      challengerPoemId: json['challengerPoemId'] as String?,
      challengeePoemId: json['challengeePoemId'] as String?,
      status: json['status'] as String,
      votesForChallenger: json['votesForChallenger'] as int? ?? 0,
      votesForChallengee: json['votesForChallengee'] as int? ?? 0,
      challengerPercent: json['challengerPercent'] as int? ?? 50,
      challengeePercent: json['challengeePercent'] as int? ?? 50,
      totalVotes: json['totalVotes'] as int? ?? 0,
      voterIds: json['voterIds'] != null
          ? List<String>.from(json['voterIds'] as List)
          : [],
      winnerId: json['winnerId'] as String?,
      endsAt: json['endsAt'] != null
          ? DateTime.parse(json['endsAt'] as String)
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  bool hasVoted(String userId) => voterIds.contains(userId);
}
